### ISSUE TITLE
Segfault when using `--iree-execution-model=inline-static`

### ISSUE THREAD

**Issue by dcaballe** [iree-org/iree#14587]
It looks like something got out of sync with the `inline-static` and `host-only` execution modes.
To repro:
```
iree-compile -iree-input-type=stablehlo --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu --iree-execution-model=inline-static ./tests/e2e/stablehlo_ops/add.mlir
```

Taking also this opportunity to ask, is there a way to lower the host function (vm) to LLVM IR and generate a binary (hopefully without going through emitC). We are looking into this to deal with some of the super tiny benchmarks that we have for CPU.

Thanks!

**Comment by benvanik** [comment]
Segfaults don't mean much without stack traces :) Please provide stack traces of failures, reproducers, etc.

No path to LLVM besides emitc for the foreseeable future - tiny models where such things matter aren't really a focus of the project at this time.

**Comment by dcaballe** [comment]
Sorry if it wasn't obvious. The .mlir file in the provided repro command is an IREE test so you can repro by just copy-pasting and running the command from the IREE root directory.

**Comment by ScottTodd** [comment]
Strange... I can repro a crash with `--iree-execution-model=inline-static` but I'm not able to see symbols or even stop at a breakpoint to debug. All I see (RelWithDebInfo, command line or via VSCode's debugger) is:
```
Exception Code: 0xC0000005
iree-compile.exe!00007ff721ec967a() (Unknown Source:0)
iree-compile.exe!00007ff71b95f9af() (Unknown Source:0)
iree-compile.exe!00007ff71b9645fa() (Unknown Source:0)
iree-compile.exe!00007ff72189b460() (Unknown Source:0)
iree-compile.exe!00007ff7218a2e17() (Unknown Source:0)
iree-compile.exe!00007ff7218a4e1d() (Unknown Source:0)
iree-compile.exe!00007ff7218a4ceb() (Unknown Source:0)
iree-compile.exe!00007ff7218a34e6() (Unknown Source:0)
iree-compile.exe!00007ff71b77ff9f() (Unknown Source:0)
iree-compile.exe!00007ff71b734d11() (Unknown Source:0)
iree-compile.exe!00007ff71b73c752() (Unknown Source:0)
iree-compile.exe!00007ff725631194() (Unknown Source:0)
kernel32.dll!00007ff965e37614() (Unknown Source:0)
ntdll.dll!00007ff967ba26b1() (Unknown Source:0)
```


Each of the other execution models (e.g. `--iree-execution-model=inline-dynamic`) works.

Also note that the command in the issue description includes a flag twice: `--iree-hal-target-backends=llvm-cpu --iree-hal-target-backends=llvm-cpu` (I can repro with and without the duplication)

**Comment by ScottTodd** [comment]
Inline static works with `--iree-hal-target-backends=vmvx-inline`, at least. We should be emitting an error somewhere instead of crashing though.

**Comment by benvanik** [comment]
inline-static only works with the vmvx-inline backend

**Comment by ScottTodd** [comment]
Yep, only working with `vmvx-inline` is expected: https://github.com/openxla/iree/blob/b41df2ff65860eb8ed0d7333075b59f86da249a2/compiler/src/iree/compiler/Pipelines/Options.h#L109-L112

Crashing on other backends could be avoided though. Maybe I can printf debug somewhere, if the debugger isn't working and crash symbols are missing...

**Comment by dcaballe** [comment]
Thanks!

> inline-static only works with the vmvx-inline backend

Any idea of what would be missing to have it working for `llvm-cpu`?

> Each of the other execution models (e.g. --iree-execution-model=inline-dynamic) works.

Does `host-only` works for you? I'm getting legalization issues:

```
./tests/e2e/stablehlo_ops/add.mlir:5:3: error: failed to legalize operation 'arith.constant' that was explicitly marked illegal                                           
```

**Comment by ScottTodd** [comment]
> > inline-static only works with the vmvx-inline backend
> 
> Any idea of what would be missing to have it working for `llvm-cpu`?

Blame layer points to https://github.com/openxla/iree/pull/9945, with some comments:

> The inline-static mode works with the vmvx-inline HAL target to produce a completely inlined VM program using the vmvx module for codegen translation. The compiler flow is nearly identical to the full HAL only when lowering out of the stream dialect we switch to the minimal hal_inline dialect and inline the codegen-produced executable contents right into the module.

It's a different computing model. What are your requirements? There might be a different configuration that supports whatever it is you are trying to do.

**Comment by ScottTodd** [comment]
Lol, added printfs and now I can get a stack trace:
```
iree-compile.exe!mlir::SymbolTable::getSymbolVisibility(mlir::Operation * symbol) Line 236 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\IR\SymbolTable.cpp:236)
[Inline Frame] iree-compile.exe!mlir::detail::SymbolOpInterfaceTrait<mlir::func::FuncOp>::getVisibility() Line 181 (d:\dev\projects\iree-build\llvm-project\tools\mlir\include\mlir\IR\SymbolInterfaces.h.inc:181)
[Inline Frame] iree-compile.exe!mlir::detail::SymbolOpInterfaceTrait<mlir::func::FuncOp>::isPublic() Line 193 (d:\dev\projects\iree-build\llvm-project\tools\mlir\include\mlir\IR\SymbolInterfaces.h.inc:193)
iree-compile.exe!mlir::iree_compiler::IREE::HAL::Inline::InlineExecutablesPass::inlineVariant(mlir::iree_compiler::IREE::HAL::ExecutableOp executableOp, mlir::iree_compiler::IREE::HAL::ExecutableVariantOp variantOp, mlir::ModuleOp targetModuleOp, llvm::DenseMap<mlir::Attribute,mlir::Attribute,llvm::DenseMapInfo<mlir::Attribute,void>,llvm::detail::DenseMapPair<mlir::Attribute,mlir::Attribute>> & exportToFuncMap, mlir::SymbolTableCollection & symbolTables) Line 135 (d:\dev\projects\iree\compiler\src\iree\compiler\Modules\HAL\Inline\Transforms\InlineExecutables.cpp:135)
iree-compile.exe!mlir::iree_compiler::IREE::HAL::Inline::InlineExecutablesPass::runOnOperation() Line 51 (d:\dev\projects\iree\compiler\src\iree\compiler\Modules\HAL\Inline\Transforms\InlineExecutables.cpp:51)
[Inline Frame] iree-compile.exe!mlir::detail::OpToOpPassAdaptor::run::__l2::<lambda_e8f8990a45bf3495636c03506b9db479>::operator()() Line 485 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\Pass\Pass.cpp:485)
iree-compile.exe!llvm::function_ref<void __cdecl(void)>::callback_fn<<lambda_e8f8990a45bf3495636c03506b9db479>>(__int64 callable) Line 45 (d:\dev\projects\iree\third_party\llvm-project\llvm\include\llvm\ADT\STLFunctionalExtras.h:45)
[Inline Frame] iree-compile.exe!llvm::function_ref<void __cdecl(void)>::operator()() Line 68 (d:\dev\projects\iree\third_party\llvm-project\llvm\include\llvm\ADT\STLFunctionalExtras.h:68)
[Inline Frame] iree-compile.exe!mlir::MLIRContext::executeAction(llvm::function_ref<void __cdecl(void)>) Line 275 (d:\dev\projects\iree\third_party\llvm-project\mlir\include\mlir\IR\MLIRContext.h:275)
iree-compile.exe!mlir::detail::OpToOpPassAdaptor::run(mlir::Pass * pass, mlir::Operation * op, mlir::AnalysisManager am, bool verifyPasses, unsigned int parentInitGeneration) Line 491 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\Pass\Pass.cpp:491)
iree-compile.exe!mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager & pm, mlir::Operation * op, mlir::AnalysisManager am, bool verifyPasses, unsigned int parentInitGeneration, mlir::PassInstrumentor * instrumentor, const mlir::PassInstrumentation::PipelineParentInfo * parentInfo) Line 551 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\Pass\Pass.cpp:551)
iree-compile.exe!mlir::PassManager::runPasses(mlir::Operation * op, mlir::AnalysisManager am) Line 860 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\Pass\Pass.cpp:860)
iree-compile.exe!mlir::PassManager::run(mlir::Operation * op) Line 839 (d:\dev\projects\iree\third_party\llvm-project\mlir\lib\Pass\Pass.cpp:839)
iree-compile.exe!mlir::iree_compiler::embed::`anonymous namespace'::Invocation::runPipeline(iree_compiler_pipeline_t pipeline) Line 763 (d:\dev\projects\iree\compiler\src\iree\compiler\API\Internal\Embed.cpp:763)
iree-compile.exe!mlir::iree_compiler::runIreecMain::__l2::<lambda>(iree_compiler_source_t * source) Line 215 (d:\dev\projects\iree\compiler\src\iree\compiler\Tools\iree_compile_lib.cc:215)
iree-compile.exe!mlir::iree_compiler::runIreecMain(int argc, char * * argv) Line 295 (d:\dev\projects\iree\compiler\src\iree\compiler\Tools\iree_compile_lib.cc:295)
[Inline Frame] iree-compile.exe!invoke_main() Line 78 (d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78)
iree-compile.exe!__scrt_common_main_seh() Line 288 (d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
kernel32.dll!00007ff965e37614() (Unknown Source:0)
ntdll.dll!00007ff967ba26b1() (Unknown Source:0)
```

**Comment by dcaballe** [comment]
> What are your requirements? 

The ultimate goal is to minimize the overhead of initializing the vm and orchestrating dispatches, ideally by creating a single dispatch and somehow compiling (and not interpreting) the vm instructions on the host side (or any other similar approach). The current situation is that we have some benchmarks where the vm initialization/dispatching is 98% of the end-to-end execution time. Of course, it's not even clear if these requirements can be met. That's why we are looking into all these execution model options.

Thanks a lot for looking into this!

**Comment by ScottTodd** [comment]
> The current situation is that we have some benchmarks where the vm initialization/dispatching is 98% of the end-to-end execution time.

I'm leaning towards +1'ing Ben's earlier comment for this

> tiny models where such things matter aren't really a focus of the project at this time.

These systems (MLIR-based compilers) are designed to take large/complex programs and distill them down to forms compatible with hardware. The hardware/target platform itself may be small/specialized or large/general purpose, but the input programs are typically complex enough that a compiler has plenty of work to do.

> ideally by creating a single dispatch and somehow compiling (and not interpreting) the vm instructions on the host side (or any other similar approach). 

This could work with emit-c or vmvx-inline... for prototyping, something built on https://github.com/openxla/iree/pull/14578 and https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/local/executable_library_test.c to call directly into an executable would avoid most of the runtime overheads.

**Comment by benvanik** [comment]
If you're seeing 98% of your execution time (how/what did you measure?) in the VM you're probably running `return 4;` or something beyond the minimum complexity we target. inline-dynamic with emitc and static linking of the executables is as small/low-overhead as you're going to get from IREE today. There's many routes to getting smaller than that but they're a very rare use case in ML that's usually better served with different approaches or with a lot more work than a few flag flips will give you - you're going to need to get _much_ more involved to poke around at that level. If a dozen jumps, some stack-local parameter shuffling, and a few indirect function calls dominate your total end-to-end latency then you're effectively in hello world territory and it entails an entirely different set of solutions from all the problems we have/are/are likely to look at. Such small cases are best solved by batching work and not by shaving off nanoseconds of individual items.

In other words, if you have a problem at the ~hundred-to-thousand CPU instruction scale with hard real-time requirements then IREE (or any generalized ML framework) is not the way to go. Vanishingly few problems demand the effort involved in meeting a hard real-time requirement. If you don't _actually_ have a hard real-time requirement then batching is what you should do regardless of what you choose to implement it.

**Comment by dcaballe** [comment]
> inline-static only works with the vmvx-inline backend

I was talking to @hcindyl yesterday. It looks like they are using `inline-static` + `llvm-cpu` on a daily basis. That was surprising to me, give the discussion above. I wonder if it's still working for them after [Scott's patch](https://github.com/openxla/iree/commit/1a0b3fd28eb72558662e191a24d9767a679c8c16). She pointed to https://github.com/AmbiML/iree-rv32-springbok/blob/main/samples/util/CMakeLists.txt#L45 and the samples in the parent folder. I might be missing some implementation details, though.

**Comment by benvanik** [comment]
That's different - that's still inline-dynamic with the static library loader - inline-static has no loader and _only_ supports vmvx.
